indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java

```diff
@@ -168,10 +169,33 @@ public static HadoopIngestionSpec updateSegmentListIfDatasourcePathSpecIsUsed(
           ingestionSpecMap,
           DatasourceIngestionSpec.class
       );
+
       List<DataSegment> segmentsList = segmentLister.getUsedSegmentsForIntervals(
           ingestionSpecObj.getDataSource(),
           ingestionSpecObj.getIntervals()
       );
+
+      if (ingestionSpecObj.getSegments() != null) {
+        //ensure that user supplied segment list matches with the segmentsList obtained from db
+        //this safety check lets users do test-n-set kind of batch delta ingestion where the delta
+        //ingestion task would only run if current state of the system is same as when they submitted
+        //the task.
+        List<DataSegment> userSuppliedSegmentsList = ingestionSpecObj.getSegments();
+
+        if (segmentsList.size() == userSuppliedSegmentsList.size()) {
+          Collections.sort(segmentsList);
```

Comment:
this assumes that the arrays returned by segmentLister and contained in the ingestion spec obj will be mutable, which might not always be the case. would be more robust to make copies of them as HashSets or TreeSets and then compare the sets for equality.